Add optional optimizer

joshdavham opened the pull request:
I've added a new `Optimizer` class that users can optionally use to compute a set of optimized `Scheduler` parameters from their `ReviewLog` objects.

Users can compute optimized parameters like so:
```python
from fsrs import ReviewLog, Optimizer, Scheduler

# load your ReviewLog objects into a list (order doesn't matter)
review_logs = [ReviewLog1, ReviewLog2, ...]

# initialize the optimizer with the review logs
optimizer = Optimizer(review_logs)

# compute a set of optimized parameters
optimal_parameters = optimizer.compute_optimal_parameters()

# initialize a new scheduler with the optimized parameters!
scheduler = Scheduler(parameters=optimal_parameters)
```

Furthermore, because the new `Optimizer` class would rely on [pytorch](https://github.com/pytorch/pytorch), the optimizer is made available as an optional depdendency and installed via:
```
pip install "fsrs[optimizer]"
```

The basic fsrs package would still be available through
```
pip install fsrs
```

and would still be pure python with no additional dependencies.

Additionally, I've seperated the unit tests into two files `test_basic.py` and `test_optimizer.py` where `test_basic.py` contains all the old scheduler unit tests and `test_optimizer.py` contains unit tests specifically for the optimizer. I've also modified our `test.yml` github workflow to run the `test_basic.py` tests for python versions 3.10-3.13 and to run the tests in `test_optimizer.py` for python versions 3.10-3.12 since pytorch currently doesn't support python 3.13.

Finally, I've also added some additional `dev` dependencies and updated the `README.md` and `CONTRIBUTING.md` to document the new `Optimizer` class.

joshdavham commented:
@L-M-Sherlock 

I'd like to do a pre-release for this update on PyPI, however, before I bump the version number, I'd like to get your opinion:

Currently, the `ReviewLog` class is as follows:
```python
class ReviewLog:

    card: Card
    rating: Rating
    review_datetime: datetime
    review_duration: int | None
```

This seems to be a valid way of representing review logs, however, I think that including the `Card` object might've been a bit overkill. Including the `Card` object and all of its class attributes like `state`, `stability` and `difficulty` isn't very helpful imo and could be removed. Storing these `Card` attributes are not necessary for the Optimizer to work and, in any case, could just be re-calculated if a user were to simply 'simulate' their previous review logs with the scheduler themselves.

I think a better `ReviewLog` class could be:
```python
class ReviewLog:

    card_id: int
    rating: Rating
    review_datetime: datetime
    review_duration: int | None
```

which is much more simple and straight-forward.

But anyways, in the context of this PR, I think we have two options:
1. Keep the `ReviewLog` class as it currently is and release this update as a minor release (e.g., `4.2.0`
or
2. Modify the `ReviewLog` class to the new definition and release this update as a major release (e.g., `5.0.0`)

I personally vote in favor of 2. as, if we are to change the `ReviewLog` class, now would be the best time to do it since users may end up creating a bunch of `ReviewLog` classes in the meantime which they would then have to update later, which could be a hassle.

What do you think?

L-M-Sherlock commented:
How will you assign card_id to the Card?

joshdavham commented:
It's currently the epoch miliseconds of when the card was created, similar to Anki. See [here](https://github.com/open-spaced-repetition/py-fsrs/blob/dd1e48e81eb57e0a041f519d28390969bfebec0d/fsrs/fsrs.py#L97-L100).

L-M-Sherlock commented on tests/review_logs_josh_1711744352250_to_1728234780857.csv:
I don't recommend tracking this file in git.

joshdavham replied:
What would you recommend doing instead? I think it's decent data to use for the tests. Are you recommending hosting it elsewhere?

joshdavham replied:
Okay so I went and did a bunch of research and apparently it's totally fine to track .csv files like this in git.

csv files are simple text files just like .py and .md files and so they are tracked efficiently and without redundancy in git, especially when they are small files.

The types of files you don't want to commit however are large binaries like images, audio or that sqlite anki database that you linked to. Those are generally recommended to get commited to [git LFS](https://git-lfs.com/).

If the csv I were adding, on the other hand were very large and was receiving tons of changes regularly, then I'd maybe potentially wanna look into hosting it somewhere else, but this particulary csv file is small enough to be fine to track with git.

L-M-Sherlock commented:
> It's currently the epoch miliseconds of when the card was created, similar to Anki. See [here](https://github.com/open-spaced-repetition/py-fsrs/blob/dd1e48e81eb57e0a041f519d28390969bfebec0d/fsrs/fsrs.py#L97-L100).

What if the application generate new cards in batch?

joshdavham commented:
> What if the application generates new cards in batch?

Great question! But to be sure, when you say generate in a batch, this is what you mean?

```python
from fsrs import Card

cards = []
for i in range(100):
    card = Card()
    cards.append(card)
```

joshdavham commented:
Okay well I just verified and a loop in python is faster than a milisecond, so in order to avoid `card_id` collisions when creating cards in a batch, we could switch to epoch microseconds.

```python
# miliseconds
from datetime import datetime, timezone

for i in range(10):
    
    timestamp = int(datetime.now(timezone.utc).timestamp() * 1000)
    print(timestamp)

"""
1737441006901
1737441006901
1737441006901
1737441006901
1737441006901
1737441006901
1737441006901
1737441006901
1737441006901
1737441006901
"""
```

```python
# microseconds
from datetime import datetime, timezone

for i in range(10):
    
    timestamp = int(datetime.now(timezone.utc).timestamp() * 1000000)
    print(timestamp)

"""
1737441087271209
1737441087271284
1737441087271293
1737441087271300
1737441087271306
1737441087271312
1737441087271317
1737441087271323
1737441087271329
1737441087271335
"""
```

I could get on this tmo.

L-M-Sherlock commented:
> Great question! But to be sure, when you say generate in a batch, this is what you mean?

At least we need to warn developer not to create cards in batch by default. Otherwise they should deal with the id stuff by themselves.

joshdavham commented:
> At least we need to warn developer not to create cards in batch by default.

Yeah we could definitely do that, but I suspect it's likely not a problem since microseconds is probably at a level of granularity where there won't be any collisions. For example, on my Apple M2 Pro chip, it looks like a loop in python is at least 5 microseconds, so unless the developers chip is 5x faster than that, there won't be a problem (unless they do parallel processing).

But if you still want me to make a note about batch processing, I can definitely do that.

L-M-Sherlock commented:
I still get collisions when I run this code:

```python
from fsrs import Card

cards = [Card() for _ in range(100)]
print(len(set([card.card_id for card in cards])))
print([card.card_id for card in cards])
```

The output:

```
87
[1737526064579865, 1737526064579871, 1737526064579872, 1737526064579874, 1737526064579875, 1737526064579876, 1737526064579877, 1737526064579878, 1737526064579878, 1737526064579879, 1737526064579880, 1737526064579881, 1737526064579882, 1737526064579883, 1737526064579884, 1737526064579885, 1737526064579885, 1737526064579886, 1737526064579887, 1737526064579888, 1737526064579889, 1737526064579890, 1737526064579891, 1737526064579892, 1737526064579893, 1737526064579893, 1737526064579894, 1737526064579895, 1737526064579896, 1737526064579897, 1737526064579898, 1737526064579898, 1737526064579899, 1737526064579900, 1737526064579901, 1737526064579902, 1737526064579903, 1737526064579904, 1737526064579905, 1737526064579905, 1737526064579906, 1737526064579907, 1737526064579908, 1737526064579909, 1737526064579910, 1737526064579911, 1737526064579911, 1737526064579912, 1737526064579913, 1737526064579914, 1737526064579915, 1737526064579915, 1737526064579916, 1737526064579917, 1737526064579918, 1737526064579919, 1737526064579920, 1737526064579921, 1737526064579921, 1737526064579922, 1737526064579923, 1737526064579924, 1737526064579925, 1737526064579926, 1737526064579927, 1737526064579928, 1737526064579929, 1737526064579929, 1737526064579930, 1737526064579931, 1737526064579932, 1737526064579933, 1737526064579934, 1737526064579934, 1737526064579935, 1737526064579936, 1737526064579937, 1737526064579938, 1737526064579941, 1737526064579942, 1737526064579943, 1737526064579944, 1737526064579944, 1737526064579945, 1737526064579946, 1737526064579947, 1737526064579948, 1737526064579956, 1737526064579957, 1737526064579958, 1737526064579958, 1737526064579959, 1737526064579960, 1737526064579961, 1737526064579962, 1737526064579963, 1737526064579964, 1737526064579965, 1737526064579965, 1737526064579966]
```

joshdavham commented:
Alright, I went and added a note in the README about how to do batch card creation while avoiding card id collisions.

I also decided to move card ids back to being milliseconds rather than microseconds so as to be consistent with anki since microseconds didn't solve the problem anyway. I also did look into whether epoch nanoseconds could be a viable solution but many operating systems don't track that level of resolution (including my os).

Also, I bumped the version to `5.0.0` for release.

L-M-Sherlock commented:
The `fsrs.py` file is too long now. Would you mind moving the optimizer part into a single file?

joshdavham commented:
> The fsrs.py file is too long now. Would you mind moving the optimizer part into a single file?

Sure thing. I went and moved `Optimizer` into its own module.

I also changed `DEFAULT_PARAMETERS` into an immutable tuple and added a short docstring to the optimizer module while I was at it.

L-M-Sherlock commented:
Could you add a progress bar to the optimizer? The test costs me near 1 minute and I don't know how much time it costs to run a epoch.

joshdavham commented:
I suppose we could. But does fsrs-rs use a progress bar anywhere?

L-M-Sherlock commented:
> I suppose we could. But does fsrs-rs use a progress bar anywhere?

Yes. FSRS-rs doesn't have a progress bar, but it could report the progress to Anki.

joshdavham commented:
Alrighty. I'll go and look into adding a progress bar tmo and I'll let you know!

Anything else while you're here? I'd like to do a release soon.

L-M-Sherlock reviewed (approved):
LGTM

joshdavham commented:
Do you still want the progress bar or am I good to pack a release tmo without it?

L-M-Sherlock commented:
Let's leave it in the next release.
